README.md
# ooutils

OO utils - Collection of Onion Omega python scripts.

## Installation

### libs

Install extra libraries that this package depends on:
``` sh
opkg install python-light power-dock2
```

### setup.py

Install this package manually via **setup.py** file:
``` sh
git clone https://github.com/kyuzom/ooutils
cd ooutils
python setup.py install
```

## Usage

### python lib

Use as an external library:
``` python
from __future__ import print_function
import time
import ooutils

# LiPo battery data
battery = ooutils.Battery()
print(battery.level_raw())
print(battery.level())
print(battery.percentage(vmax=4.12, timeout=0.8))

# LiPo battery asynchronous data
battery = ooutils.ABattery()
id0 = battery.level_raw()
id1 = battery.level()
id2 = battery.percentage(vmax=4.12, timeout=0.8)
print(battery.wait(id0, timeout=0.1))
print(battery.wait(id2))
time.sleep(3)
print(battery.get(id0))
print(battery.get(id1))
print(battery.terminate(id2))
print(battery.get(id2))
```

## License

ooutils is MIT licensed. See the included [LICENSE](LICENSE) file.


CHANGELOG.md
# Change Log

Changes to gpydem package.

## [0.0.3] - 2021-12-05

 - LiPo battery asynchronous working mode: wait, terminate

## [0.0.2] - 2021-12-05

 - LiPo battery Timeout support

## [0.0.1] - 2021-12-03

 - LiPo battery support: V level, percentage


ooutils/battery.py
#!/usr/bin/env python
'''
Onion Omega LiPo battery support.
'''
import subprocess
import threading
import re
try:
	from Queue import Queue
except (ImportError, ModuleNotFoundError):
	from queue import Queue

class Battery(object):
	'''
	LiPo battery object
	'''
	BATTERY_CLI_EXE = "power-dock2"
	BATTERY_RGX_PAT = "(\d+\.\d+ V)|(\d+ V)"

	def __init__(self):
		self._ps = None

	def level_raw(self, timeout=1.0):
		'''
		Get battery level in raw format (str)
		@param timeout: [float] Timeout for battery level reading
		@return [str] Battery level in format '%d V'
		'''
		def access_battery_level(msgq):
			self._ps = subprocess.Popen([Battery.BATTERY_CLI_EXE], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
			out, err = self._ps.communicate()
			msgq.put((self._ps.returncode, out, err), block=True, timeout=0.1)
		msgq = Queue(maxsize=1)
		psthr = threading.Thread(target=access_battery_level, args=(msgq,))
		psthr.start()
		psthr.join(timeout=timeout)
		if self._ps is None:
			raise RuntimeError("OO Battery level, Timed out.")
		elif (self._ps.poll() is None) or psthr.is_alive():
			self._ps.terminate()
			psthr.join()
			self._ps = None
			raise RuntimeError("OO Battery level, Timed out.")
		self._ps = None
		rc, out, err = msgq.get(block=True, timeout=0.1)
		msgq.task_done()
		msgq.join()
		if rc != 0:
			raise RuntimeError("OO Battery level, rc: {0}, error: {1}".format(rc, err))
		match = re.search(Battery.BATTERY_RGX_PAT, out.strip())
		if match is None:
			raise RuntimeError("OO Battery level, Unknown str format: {0}".format(out))
		return match.group()

	def level(self, timeout=1.0):
		'''
		Get battery level in [V] (float)
		@param timeout: [float] Timeout for battery level reading
		@return [float] Battery level in [V]
		'''
		return float(self.level_raw(timeout=timeout)[:-2])

	def percentage(self, vmax=4.2, vmin=3.5, timeout=1.0):
		'''
		Get battery level percentage in [%] (float)
		@param vmax: [float] Battery maximum level in [V]
		@param vmin: [float] Battery minimum level in [V]
		@param timeout: [float] Timeout for battery level reading
		@return [float] Battery level in [%]
		'''
		vactual = self.level(timeout=timeout)
		return float(((vactual-vmin)/(vmax-vmin))*100)

if __name__ == '__main__':
	import sys
	battery = Battery()
	if sys.version_info[0] == 2:
		print battery.level_raw()
	else:
		print(battery.level_raw())


ooutils/oobattery.py
#!/usr/bin/env python
'''
Onion Omega LiPo battery support.
'''
import subprocess
import re

BATTERY_CLI_EXE = "power-dock2"
BATTERY_RGX_PAT = "(\d+\.\d+ V)|(\d+ V)"

def battery_level_raw():
	'''
	Return with battery level in raw format (str)
	'''
	ps = subprocess.Popen(BATTERY_CLI_EXE, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
	out, err = ps.communicate()
	if ps.returncode != 0:
		raise RuntimeError("OO battery_level, error: ", err)
	match = re.search(BATTERY_RGX_PAT, out.strip())
	if match is None:
		raise RuntimeError("OO battery_level, Could not retrieve information.")
	return match.group()

def battery_level():
	'''
	Return with battery level in [V] (float)
	'''
	return float(battery_level_raw()[:-2])

def battery_percentage(vmax=4.2, vmin=3.5):
	'''
	Return with battery level percentage in [%] (float)
	'''
	vactual = battery_level()
	return float(((vactual-vmin)/(vmax-vmin))*100)


ooutils/abattery.py
#!/usr/bin/env python
'''
Onion Omega LiPo battery asynchronous support.

NOTE! This is not the async python lib, just enables asynchronous battery level access.
'''
from __future__ import print_function
import uuid
import threading
try:
	from .battery import Battery
except ValueError:
	from battery import Battery

class ABattery(object):
	'''
	LiPo battery asynchronous object
	'''
	OP_NOTHING = 0
	OP_ONGOING = 1
	OP_SUCCESS = 2
	OP_FAILURE = 3

	def __init__(self):
		self._results = dict()

	def get(self, id):
		'''
		Get result identified by parameter id
		@param id: [str] Operation unique identifier
		@return [tuple] Result of the requested operation (operation code, result value)
		'''
		if id in self._results:
			return (self._results[id]['opcode'], self._results[id]['result'])
		else:
			return (ABattery.OP_NOTHING, None)

	def wait(self, id, timeout=1.0):
		'''
		Wait for operation to finish identified by parameter id
		@param id: [str] Operation unique identifier
        @param timeout: [float] Timeout for waiting
		@return [tuple] Result of the requested operation (operation code, result value)
		'''
		if id in self._results:
			self._results[id]['worker'].join(timeout=timeout)
			return (self._results[id]['opcode'], self._results[id]['result'])
		else:
			return (ABattery.OP_NOTHING, None)

	def terminate(self, id):
		'''
		Terminate operation identified by parameter id
		@param id: [str] Operation unique identifier
		@return [tuple] Result of the requested operation (operation code, result value)
		'''
		if id in self._results:
			battery = self._results[id]['object']
			opcode = self._results[id]['opcode']
			result = self._results[id]['result']
			if battery._ps is not None:
				battery._ps.terminate()
			self._results[id]['worker'].join()
			del self._results[id]
			return (opcode, result)
		else:
			return (ABattery.OP_NOTHING, None)

	def _battery_async(self, id, *args):
		'''
		Measure battery level in the background
		@param id: [str] Operation unique identifier
		'''
		battery = self._results[id]['object']
		try:
			value = getattr(battery, self._results[id]['method'])(*args)
		except Exception as e:
			self._results[id]['result'] = str(e)
			self._results[id]['opcode'] = ABattery.OP_FAILURE
		else:
			self._results[id]['result'] = value
			self._results[id]['opcode'] = ABattery.OP_SUCCESS

	def level_raw(self, timeout=1.0):
		'''
		Get battery level in raw format in the background (str)
		@param timeout: [float] Timeout for battery level reading
		@return [str] Operation identifier
		'''
		id = str(uuid.uuid4())
		thr = threading.Thread(target=self._battery_async, args=(id, timeout))
		self._results[id] = {
			'opcode': ABattery.OP_ONGOING,
			'object': Battery(),
			'method': 'level_raw',
			'worker': thr,
			'result': None
		}
		thr.start()
		return id

	def level(self, timeout=1.0):
		'''
		Get battery level in [V] in the background (float)
		@param timeout: [float] Timeout for battery level reading
		@return [str] Operation identifier
		'''
		id = str(uuid.uuid4())
		thr = threading.Thread(target=self._battery_async, args=(id, timeout))
		self._results[id] = {
			'opcode': ABattery.OP_ONGOING,
			'object': Battery(),
			'method': 'level',
			'worker': thr,
			'result': None
		}
		thr.start()
		return id

	def percentage(self, vmax=4.2, vmin=3.5, timeout=1.0):
		'''
		Get battery level percentage in [%] in the background (float)
		@param vmax: [float] Battery maximum level in [V]
		@param vmin: [float] Battery minimum level in [V]
		@param timeout: [float] Timeout for battery level reading
		@return [str] Operation identifier
		'''
		id = str(uuid.uuid4())
		thr = threading.Thread(target=self._battery_async, args=(id, vmax, vmin, timeout))
		self._results[id] = {
			'opcode': ABattery.OP_ONGOING,
			'object': Battery(),
			'method': 'percentage',
			'worker': thr,
			'result': None
		}
		thr.start()
		return id

if __name__ == '__main__':
	#from __future__ import print_function
	import time
	battery = ABattery()
	id0 = battery.level_raw()
	id1 = battery.level()
	id2 = battery.percentage(vmax=4.12, timeout=0.8)
	print(battery.wait(id0, timeout=0.1))
	print(battery.wait(id2))
	time.sleep(3)
	print(battery.get(id0))
	print(battery.get(id1))
	print(battery.terminate(id2))
	print(battery.get(id2))
